iree/compiler/Dialect/Flow/Transforms/MatmulToMMT4dConversion.cpp

```diff
@@ -0,0 +1,242 @@
+
+    rewriter.replaceOp(matmulOp, ArrayRef<Value>{result});
+
+    return success();
+  }
+
+ private:
+  int M0Size;
+  int N0Size;
+  int K0Size;
+};
+
+/// Canonicalizes [linalg.init_tensor -> linalg.fill -> linalg.generic] ->
+/// [linalg.init_tensor -> linalg.fill] where linalg.generic does only copy e.g
+/// a transpose.
+struct FoldFillGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (genericOp.getNumInputs() != 1) return failure();
+    if (genericOp.getNumOutputs() != 1) return failure();
+
+    // Check linalg.generic does have copy only semantics.
+    if (genericOp.getNumParallelLoops() != genericOp.getNumLoops())
+      return failure();
+    auto results =
+        llvm::to_vector<4>(genericOp.getBody()->getOps<linalg::YieldOp>());
+    if (results.size() != 1) return failure();
+    if (results[0].values().size() != 1) return failure();
+    auto blockArgument = results[0].values()[0].dyn_cast<BlockArgument>();
+    if (!blockArgument || blockArgument.getArgNumber() != 0) return failure();
+
+    auto input = genericOp.inputs()[0];
+
+    auto outputType =
+        genericOp.outputs()[0].getType().dyn_cast<RankedTensorType>();
+
+    // TODO: To enable dynamic shapes we need to apply the same permutation on
+    // init tensor sizes.
+    if (!outputType || !outputType.hasStaticShape()) return failure();
+
+    auto fillOp = dyn_cast<linalg::FillOp>(input.getDefiningOp());
+    if (!fillOp) return failure();
+
+    auto initTensorOp =
+        dyn_cast<linalg::InitTensorOp>(fillOp.output().getDefiningOp());
+
+    if (!initTensorOp) return failure();
+
+    auto loc = genericOp.getLoc();
+    Value newInitTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, outputType.getShape(), outputType.getElementType());
+    rewriter.replaceOpWithNewOp<linalg::FillOp>(genericOp, fillOp.value(),
+                                                newInitTensor);
+
+    return success();
+  }
+};
+
+class ConvertLinalgMatmulOpToLinalgMMT4dPass
+    : public ConvertMatmulToMMT4dBase<ConvertLinalgMatmulOpToLinalgMMT4dPass> {
+ public:
+  ConvertLinalgMatmulOpToLinalgMMT4dPass(int M0, int N0, int K0)
+      : M0Size(M0), N0Size(K0), K0Size(N0) {}
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect>();
+  }
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    // Main pattern.
+    {
+      OwningRewritePatternList patterns(&getContext());
+      patterns.insert<LinalgMatmulOpToLinalgMMT4dOpPattern>(context, M0Size,
+                                                            N0Size, K0Size);
+      (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
+    }
+    // Canonicalization.
```

Comment:
Add a TODO somewhere here to move this upstream. There is nothing IREE specific here.